gnrc_sixlowpan_nd: register new addresses for AAC

authmillenon opened the pull request:
New addresses are are not registered for auto-address configuration in a LoWPAN. This fixes that.

OlegHahm commented on sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c:
```diff
@@ -551,7 +562,8 @@ static int _create_candidate_set(gnrc_ipv6_netif_t *iface, const ipv6_addr_t *ds
          *  be included in a candidate set."
          */
-        if (ipv6_addr_is_multicast(&(iter->addr)) ||
+        if ((iter->flags & GNRC_IPV6_NETIF_ADDR_FLAGS_TENTATIVE) ||
```
Is this specified somewhere?

authmillenon replied:
Well a tentative address is not really assigned to the interface until the registration is complete (see https://tools.ietf.org/html/rfc4862), so it needs to be skipped here... speaking of… I forgot to check for it on receival.

OlegHahm commented:
ACK - please squash

authmillenon commented:
Squashed

authmillenon commented:
Fixed error exposed by travis and squashed immediately:

```
$ git diff HEAD@{1}
diff --git a/sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c b/sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_ne
index e00e7a1..54af5a5 100644
--- a/sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c
+++ b/sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c
@@ -92,7 +92,7 @@ static ipv6_addr_t *_add_addr_to_entry(gnrc_ipv6_netif_t *entry, const ipv6_addr
         ipv6_addr_t *router = gnrc_ndp_internal_default_router();
         if (router != NULL) {
             tmp_addr->flags |= GNRC_IPV6_NETIF_ADDR_FLAGS_TENTATIVE;
-            gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, &all_routers);
+            gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, &router);
         }
         /* otherwise there is no default router to register to */
     }
```

OlegHahm commented:
Sending NS to the default router instead of multicasting, right?

authmillenon commented:
Yes. This was previously also that way. I just forgot to rename the variable here.

OlegHahm commented on sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c:
```diff
@@ -87,6 +87,17 @@ static ipv6_addr_t *_add_addr_to_entry(gnrc_ipv6_netif_t *entry, const ipv6_addr
+        ipv6_addr_t *router = gnrc_ndp_internal_default_router();
+        if (router != NULL) {
+            tmp_addr->flags |= GNRC_IPV6_NETIF_ADDR_FLAGS_TENTATIVE;
+            gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, &router);
```
`s/&router/router/`

authmillenon replied:
Addressed and squashed already

authmillenon commented:
Adapted for changes discussed offline

OlegHahm commented on sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c:
```diff
@@ -87,6 +87,17 @@ static ipv6_addr_t *_add_addr_to_entry(gnrc_ipv6_netif_t *entry, const ipv6_addr
+        ipv6_addr_t *router = gnrc_ndp_internal_default_router();
+        if (router != NULL) {
+            tmp_addr->flags |= GNRC_IPV6_NETIF_ADDR_FLAGS_TENTATIVE;
+            gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, router);
```
You need to unlock the mutex first and lock it afterwards again.

OlegHahm replied:
and `s/gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, router)/gnrc_ndp_internal_send_nbr_sol(entry->pid, &tmp_addr->addr, router, router)`

authmillenon replied:
Thought I did, but well.

OlegHahm commented on sys/net/gnrc/network_layer/ndp/gnrc_ndp.c:
```diff
@@ -609,9 +610,13 @@ void gnrc_ndp_rtr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt, ipv6_hdr_t
+                                       &nc_entry->ipv6_addr);
+        vtimer_remove(&nc_entry->nbr_sol_timer);
+        vtimer_set_msg(&nc_entry->nbr_sol_timer, t, gnrc_ipv6_pid, GNRC_NDP_MSG_NBR_SOL_RETRANS,
+                       nc_entry);
```
This seems to revert https://github.com/RIOT-OS/RIOT/pull/3874

OlegHahm replied:
see https://github.com/authmillenon/RIOT/pull/22

authmillenon replied:
Thanks.

OlegHahm replied:
Something went wrong again...

authmillenon replied:
WTF?

authmillenon replied:
Will fix.

authmillenon replied:
Fixed and squashed immediately.

OlegHahm commented on sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c:
```diff
@@ -87,6 +87,19 @@ static ipv6_addr_t *_add_addr_to_entry(gnrc_ipv6_netif_t *entry, const ipv6_addr
+    if (entry->flags & GNRC_IPV6_NETIF_FLAGS_SIXLOWPAN) {
+        ipv6_addr_t *router = gnrc_ndp_internal_default_router();
+        if (router != NULL) {
+            tmp_addr->flags |= GNRC_IPV6_NETIF_ADDR_FLAGS_TENTATIVE;
```
If you never unset this, it will stay TENTATIVE forever...

authmillenon replied:
see 6e8f7c4

OlegHahm commented:
Just a git technical question: why are https://github.com/authmillenon/RIOT/commit/6e8f7c455440e2715f55956b50dc6d3ffd0780f9 and https://github.com/authmillenon/RIOT/commit/91e0a5f5e25bf5bd73a9dbde0ae30860073980ff showing the same diff for the same file?

OlegHahm commented:
Pleae squash.

authmillenon commented:
> Just a git technical question: why are authmillenon@6e8f7c4 and authmillenon@91e0a5f showing the same diff for the same file?

Because they both provide the same change. And since the diff (or rather the resulting blob) is identical, the merge commit at ce07d9e has no problem merging them. It's basically alternative history converging at the merge commit ;-)

authmillenon commented:
Squashed

OlegHahm commented:
ACK
